<a href="https://colab.research.google.com/github/Lizeth058/BIT_BootcampIA/blob/main/Paratrabajar_Colombia_bankrupy_prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Paso 1. Librerias y

In [4]:

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, classification_report, recall_score, roc_auc_score, f1_score
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.impute import SimpleImputer
from imblearn.under_sampling import TomekLinks
from sklearn.tree import plot_tree
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from google.colab import drive
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler, Normalizer
from sklearn.pipeline import Pipeline
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score, StratifiedKFold



In [10]:
from google.colab import drive # Conectanonos con google drive
drive.mount('/content/drive')
import os    # la librería necesaria para la ruta de la carpeta
os.chdir('/content/drive/MyDrive/Colab Notebooks/bootcamp IA/RETO')
#mostar lista de archivos
os.listdir()

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


['example_submission.csv',
 'Untitled0.ipynb',
 'arbol.png',
 'modelo.png',
 'predictions.csv',
 'predictions2.csv',
 'test.csv',
 'train.csv']

# Componente 1. Comprensión de los datos
## Paso 1. Recopilación de los datos iniciales

In [6]:
df_train = pd.read_csv('train.csv')
df_test = pd.read_csv('test.csv') # Se usa para ejecutar la evolución

In [7]:
df = pd.read_csv('train.csv')
df.head()

index  Ganancia bruta  Ganancia (pérdida)  \
0   7620    3.218247e+05        1.322028e+06   
1  17602    1.665600e+05       -1.603930e+05   
2  18421    2.387397e+06        3.805740e+05   
3    779    5.214416e+06        7.228300e+04   
4   6936    2.271980e+05        7.254877e+05   

   Ingresos de actividades ordinarias  Costo de ventas  Patrimonio total  \
0                        3.218247e+05              NaN      9.147743e+06   
1                        1.665600e+05              NaN      3.594330e+05   
2                        3.686907e+06        1299510.0      2.421259e+06   
3                        6.610757e+07       60893155.0      5.015903e+06   
4                        2.271980e+05              NaN      8.062667e+06   

   Total pasivos  Total de activos  Ganancias acumuladas  \
0   9.729800e+04      9.245041e+06             2610498.0   
1   6.489900e+04      4.243320e+05             -160393.0   
2   2.945965e+06      5.367224e+06              461980.0   
3   2.482254e+07      2.983844e+07             4415903.0   
4   2.770852e+06      1.083352e+07             4585023.0   

   Pasivos corrientes totales  Activos corrientes totales Sector  event  
0                9.729800e+04                   3207644.0      L    0.0  
1                6.489900e+04                    193787.0      G    0.0  
2                1.615549e+06                   3364490.0      J    0.0  
3                2.261366e+07                  22996533.0      G    1.0  
4                3.307467e+04                    257324.0      L    0.0

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14097 entries, 0 to 14096
Data columns (total 13 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   index                               14097 non-null  int64  
 1   Ganancia bruta                      14097 non-null  float64
 2   Ganancia (pérdida)                  14097 non-null  float64
 3   Ingresos de actividades ordinarias  14097 non-null  float64
 4   Costo de ventas                     12335 non-null  float64
 5   Patrimonio total                    14097 non-null  float64
 6   Total pasivos                       14097 non-null  float64
 7   Total de activos                    14097 non-null  float64
 8   Ganancias acumuladas                14061 non-null  float64
 9   Pasivos corrientes totales          14059 non-null  float64
 10  Activos corrientes totales          14092 non-null  float64
 11  Sector                              14097

## Paso 2. Descripción de datos
  ¿Cuántos atributos, cuantos registros?

   13 atributos, 14097 registros

  ¿Cual es el tipo de datos de los atributos?

   1 int64, 1 object 11 float64
   
## Paso 3.Exploración de datos

## Paso 4. Verificación de la calidad

 Nucleos, man , ouilers
   

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14097 entries, 0 to 14096
Data columns (total 13 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   index                               14097 non-null  int64  
 1   Ganancia bruta                      14097 non-null  float64
 2   Ganancia (pérdida)                  14097 non-null  float64
 3   Ingresos de actividades ordinarias  14097 non-null  float64
 4   Costo de ventas                     12335 non-null  float64
 5   Patrimonio total                    14097 non-null  float64
 6   Total pasivos                       14097 non-null  float64
 7   Total de activos                    14097 non-null  float64
 8   Ganancias acumuladas                14061 non-null  float64
 9   Pasivos corrientes totales          14059 non-null  float64
 10  Activos corrientes totales          14092 non-null  float64
 11  Sector                              14097

In [41]:

from sklearn.model_selection import train_test_split, cross_validate
from sklearn.preprocessing import MinMaxScaler, StandardScaler, Normalizer
from sklearn.pipeline import Pipeline
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix, roc_auc_score, classification_report
import pandas as pd
from sklearn.model_selection import cross_validate
from sklearn.preprocessing import MinMaxScaler, StandardScaler, Normalizer
from sklearn.pipeline import Pipeline
from sklearn.tree import DecisionTreeClassifier
import numpy as np



In [55]:

# Imputar los valores faltantes con la media de cada columna
for col in ["Costo de ventas", "Ganancias acumuladas", "Pasivos corrientes totales", "Activos corrientes totales"]:
    train_df[col] = train_df[col].fillna(train_df[col].mean())
    test_df[col] = test_df[col].fillna(test_df[col].mean())

# Codificar la columna `Sector` usando one-hot encoding
X_train = pd.get_dummies(train_df, columns=["Sector"])
X_test = pd.get_dummies(test_df, columns=["Sector"])

# Extraer la variable objetivo
y_train = train_df["event"]

# Eliminar columnas innecesarias
X_train.drop(columns=["Costo de ventas", "index", "event"], inplace=True)
X_test.drop(columns=["Costo de ventas", "index"], inplace=True)

In [56]:

# Definir los pipelines
pipelines = {
    "Escalado Min-Max": Pipeline([
        ("escalador", MinMaxScaler()),
        ("clasificador", DecisionTreeClassifier(random_state=42))
    ]),
    "Escalado Estándar": Pipeline([
        ("escalador", StandardScaler()),
        ("clasificador", DecisionTreeClassifier(random_state=42))
    ]),
    "Normalización L2": Pipeline([
        ("escalador", Normalizer(norm="l2")),
        ("clasificador", DecisionTreeClassifier(random_state=42))
    ])
}

# Crear un diccionario para almacenar las puntuaciones de validación cruzada
puntuaciones = {}

# Definir múltiples métricas de evaluación
metricas_evaluacion = {
    'f1': 'f1',
    'exactitud': 'accuracy',
    'sensibilidad': 'recall',
    'AUC': 'roc_auc'
}

# Iterar sobre los pipelines y calcular las puntuaciones de validación cruzada
for nombre, pipeline in pipelines.items():
    resultados_cv = cross_validate(pipeline, X_train, y_train, cv=5, scoring=metricas_evaluacion, return_train_score=False)

    # Almacenar las puntuaciones medias para cada métrica
    puntuaciones[nombre] = {metrica: np.mean(resultados_cv[f'test_{metrica}']) for metrica in metricas_evaluacion}
    print(f"{nombre} - Puntuaciones de validación cruzada:")
    for metrica, valor in puntuaciones[nombre].items():
        print(f"   {metrica.capitalize()}: {valor:.10f}")




Escalado Min-Max - Puntuaciones de validación cruzada:
   F1: 0.0675587326
   Exactitud: 0.9650281147
   Sensibilidad: 0.0727346939
   Auc: 0.5268366610
Escalado Estándar - Puntuaciones de validación cruzada:
   F1: 0.0678320170
   Exactitud: 0.9650279889
   Sensibilidad: 0.0727346939
   Auc: 0.5268366610
Normalización L2 - Puntuaciones de validación cruzada:
   F1: 0.0802328279
   Exactitud: 0.9674399402
   Sensibilidad: 0.0808979592
   Auc: 0.5320735283


In [54]:

# Convertir el diccionario de puntuaciones en un DataFrame
puntuaciones_df = pd.DataFrame(puntuaciones).T

# Calcular el promedio de las métricas para cada pipeline y agregarlo al DataFrame
puntuaciones_df['puntuacion_promedio'] = puntuaciones_df.mean(axis=1)

# Seleccionar el pipeline con la puntuación promedio más alta
nombre_mejor_pipeline = puntuaciones_df['puntuacion_promedio'].idxmax()
mejor_pipeline = pipelines[nombre_mejor_pipeline]

# Entrenar el mejor pipeline en todo el conjunto de entrenamiento
mejor_pipeline.fit(X_train, y_train)

# Generar predicciones en el conjunto de prueba
y_pred = mejor_pipeline.predict(X_test)


In [53]:
# Convertir el diccionario de predicciones en un DataFrame
df_predictions = pd.DataFrame(predictions)

# Imprimir las primeras 5 filas de df_predictions
print("\nPrimeras 5 filas de df_predictions:")
print(df_predictions.head().to_markdown(index=False, numalign="left", stralign="left"))


Primeras 5 filas de df_predictions:
| index   | prediction   |
|:--------|:-------------|
| 3670    | 0            |
| 6916    | 0            |
| 12849   | 0            |
| 10397   | 0            |
| 16597   | 0            |


In [57]:
# Generar predicciones de probabilidad en el conjunto de prueba
y_pred_sigmoide = mejor_pipeline.predict_proba(X_test)[:, 1]

# Convertir las probabilidades a predicciones binarias
y_pred_binary = (y_pred_sigmoide > 0.5).astype(int)

# Crear un DataFrame con las predicciones binarias
y_pred_binary_df = pd.DataFrame(y_pred_binary, columns=["y_pred"])

In [46]:
df_index = df_test['index']

In [48]:
y_pred

y_pred
0          0
1          0
2          0
3          0
4          0
...      ...
6037       0
6038       0
6039       0
6040       0
6041       0

[6042 rows x 1 columns]

In [49]:
# Crear un diccionario con las predicciones
predictions = {
    'index': test_df['index'],
    'prediction': y_pred_binary_df['y_pred']
}


In [50]:
df_predictions.to_csv('predictions2.csv', index=False)